lib/grandpa/message_handler.go

```diff
@@ -149,31 +198,46 @@ func (h *MessageHandler) handleCommitMessage(msg *CommitMessage) error {
 	return nil
 }
 
-func (h *MessageHandler) handleCatchUpRequest(msg *CatchUpRequest) (*ConsensusMessage, error) {
+func (h *MessageHandler) handleCatchUpRequest(msg *CatchUpRequest, from peer.ID) error {
 	if !h.grandpa.authority {
-		return nil, nil //nolint:nilnil
+		return nil
 	}
 
-	logger.Debugf("received catch up request for round %d and set id %d",
-		msg.Round, msg.SetID)
+	logger.Debugf("received catch up request for round %d and set id %d, from %s",
+		msg.Round, msg.SetID, from)
+
+	logger.Debugf("Our latest round is %d", h.grandpa.state.round)
 
 	if msg.SetID != h.grandpa.state.setID {
-		return nil, ErrSetIDMismatch
+		return ErrSetIDMismatch
 	}
 
-	if msg.Round >= h.grandpa.state.round {
-		return nil, ErrInvalidCatchUpRound
+	if msg.Round > h.grandpa.state.round {
+		return ErrInvalidCatchUpRound
 	}
 
-	resp, err := h.grandpa.newCatchUpResponse(msg.Round, msg.SetID)
+	// We don't necessarily have to reply with the round asked in the request, we can reply
+	// with our latest round.
+	resp, err := h.grandpa.newCatchUpResponse(h.grandpa.state.round, h.grandpa.state.setID)
 	if err != nil {
-		return nil, err
+		return err
+	}
+
+	cm, err := resp.ToConsensusMessage()
+	if err != nil {
+		return err
+	}
+
+	err = h.grandpa.network.SendMessage(from, cm)
+	if err != nil {
+		return err
 	}
```

Comment:
ah i see you are sending the message here, feel free to change the `handleMessage` return params then